enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,801 @@
+
+When adding a ControlPlaneMachineSet to a new cluster, the end user will need to define the `ControlPlaneMachineSet`
+resource by copying the existing Control Plane Machine ProviderSpecs.
+Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
+field within the ControlPlaneMachineSet spec.
+
+To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook
+that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.
+
+If no Control Plane Machines exist, or they are in a non-Running state, the operator will report degraded until this
+issue is resolved. This creates a dependency for the `ControlPlaneMachineSet` operator on the Machine API. It will be
+required to run at a higher run-level than Machine API.
+
+In UPI or misconfigured clusters, a user adding a `ControlPlaneMachineSet` will result in a degraded cluster.
+Users will need to remove the invalid ControlPlaneMachineSet resource to restore their cluster to a healthy state.
+
+We do not recommend that UPI users attempt to adopt their control plane instances into Machine API due to the high
+likelihood that they cannot create an accurate configuration to replicate the original control plane instances. This
+limitation also limits the `ControlPlaneMachineSet` operator to an IPI only operator.
+
+#### Naming of Control Plane Machines owned by the ControlPlaneMachineSet
+
+In an IPI OpenShift cluster, the Control Plane Machines are named after the cluster ID followed by an index, for example
+`my-openshift-cluster-abcde-master-0`. As we will need to create additional Machines during replacement operations, we
+cannot reuse the names of the existing Machines. Instead we will use the `generateName` field within the metadata to
+create a random suffix for the new Machines, replacing the index of current Machines. For example, Machines created
+by a `ControlPlaneMachineSet` may have a name such as `my-openshift-cluster-abcde-master-fghij`.
+
+As users can today replace their Machines with differently named Machines (eg as part of a recovery process), we do not
+expect other components to be relying on the index of the Machines for any function within OpenShift, but we may
+discover during testing something that is currently unknown to us. If we discover issues with other components relying
+on a particular naming scheme for the Control Plane Machines, we will need to solve this issue within the other
+component.
+
+#### Delivery via the core OpenShift payload
+
+This new operator will form part of the core OpenShift release payload for clusters installed on/upgrading to OpenShift
+4.11. While there is a movement within OpenShift to stop adding new components to the release payload, we believe that
+this new operator should be added to the core for the following reasons:
```

Comment:
i could see this being a candidate for day0 OLM operators in the future, but yes for now i think you make a good case for it being in the payload.

I do think it would be reasonable to have it be an optional capability, however, per https://issues.redhat.com/browse/OCPPLAN-7589 and https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md

Reply:
Yep, this seems reasonable, added a note and a link to the component-selection enhancement saying this can be made optional via that mechanism